src/ast.rs

```diff
@@ -88,6 +112,23 @@ impl PrettyPrinting for Stmt {
                 s + ")"
             }
             Stmt::Expression { expression } => format!("(; {})", expression.print()),
+            Stmt::Function {
+                name,
+                parameters,
+                body,
+            } => {
+                let mut s = format!("({}", name);
+                for parameter in parameters {
+                    s += " ";
+                    s += &parameter.print();
+                }
+                s += "(body";
+                for statement in body {
+                    s += " ";
+                    s += &statement.print();
+                }
+                s + "))"
```

Comment:
Here's an alternative implementation idea:

```rust
format!(
    "({name} {parameters}(body {statements}))",
    name = name,
    parameters = parameters.iter().map(PrettyPrinting::print).collect::<Vec<_>>().join(" "),
    statements = body.iter().map(PrettyPrinting::print).collect::<Vec<_>>().join(" "),
)
```

I think it's equivalent, although the `iter().map().collect().join()` is not the most pretty. You could also implement `PrettyPrint` for `Vec<T: PrettyPrint>`:

```rust
impl<T> PrettyPrinting for Vec<T>
where
    T: PrettyPrinting,
{
    fn print(&self) -> String {
        self.iter()
            .map(PrettyPrinting::print)
            .collect::<Vec<_>>()
            .join(" ")
    }
}
```

which would make the usage:

```rust
format!(
    "({name} {parameters}(body {statements}))",
    name = name,
    parameters = parameters.print(),
    statements = body.print(),
)
```

Just spitballing alternative approaches.. feel free to ignore :)

Reply:
this is so much cleaner! thanks for the tip!